TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/BlockLibrary/VirtualTexturePropertyBlockBuilder.cs

```diff
@@ -0,0 +1,153 @@
+                PropertyAttribute = PropertyAttribute,
+                ExtraAttributes = attributes,
+                ExtraBlockGenerationCallback = (blockBuilder, propData) => { sampleFunction = CreateSampleFunction(container, blockBuilder, LayerToSample); },
+                OutputsAssignmentCallback = (builder, propData) =>
+                {
+                    builder.AddLine($"float2 uv = float2(0, 0);");
+                    builder.AddLine($"float4 sampleOut = 0;");
+                    builder.CallFunction(sampleFunction, "uv", $"inputs.{FieldName}", "sampleOut");
+                    builder.AddLine($"outputs.BaseColor = sampleOut.xyz;");
+                    builder.AddLine($"outputs.Alpha = sampleOut.w;");
+                }
+            };
+            var builder = BlockBuilderUtilities.CreateSimplePropertyBlockBuilder(container, BlockName, propData);
+            return builder.Build();
+        }
+
+        static ShaderType GetVirtualTextureType(ShaderContainer container)
+        {
+            return container.GetType(VirtualTexturePropertyTypeName);
+        }
+
+        ShaderFunction CreateSampleFunction(ShaderContainer container, Block.Builder blockBuilder, int layer)
+        {
+            var uvName = "uv";
+            var propName = "vtProperty";
+            var layerOutName = "layerOut";
+            var virtualTexturePropertyType = GetVirtualTextureType(container);
+            var functionBuilder = new ShaderFunction.Builder(blockBuilder, $"SamplerLayer{layer}", container._float4);
+            functionBuilder.AddParameter(new FunctionParameter.Builder(container, uvName, container._float2, true, false).Build());
+            functionBuilder.AddParameter(new FunctionParameter.Builder(container, propName, virtualTexturePropertyType, true, false).Build());
+            functionBuilder.AddParameter(new FunctionParameter.Builder(container, layerOutName, container._float4, false, true).Build());
+
+            functionBuilder.AddLine("VtInputParameters vtParams;");
+            functionBuilder.AddLine($"vtParams.uv = {uvName};");
+            functionBuilder.AddLine("vtParams.lodOrOffset = 0.0f;");
+            functionBuilder.AddLine("vtParams.dx = 0.0f;");
+            functionBuilder.AddLine("vtParams.dy = 0.0f;");
+            functionBuilder.AddLine("vtParams.addressMode = VtAddressMode_Wrap;");
+            functionBuilder.AddLine("vtParams.filterMode = VtFilter_Anisotropic;");
+            functionBuilder.AddLine("vtParams.levelMode = VtLevel_Automatic;");
+            functionBuilder.AddLine("vtParams.uvMode = VtUvSpace_Regular;");
+            functionBuilder.AddLine("vtParams.sampleQuality = VtSampleQuality_High;");
+            functionBuilder.AddLine("vtParams.enableGlobalMipBias = 1;");
+            functionBuilder.AddLine("#if defined(SHADER_STAGE_RAY_TRACING)");
+            functionBuilder.AddLine("if (vtParams.levelMode == VtLevel_Automatic || vtParams.levelMode == VtLevel_Bias)");
```

Comment:
Do we want to indent the code here?

Reply:
Good point. Must've missed then when pulling the code out.